### DISCUSSION TITLE
Import: Performance on large library

### DISCUSSION THREAD

**Discussion by jvallery** [photoprism/photoprism#774] [Q&A] [answered]
I have a Ryzen 7 based Ubuntu 20.04 machine with all NVMe storage, 10GbE, and 32GiB of memory dedicated to Photoprism. I have a large library ~(300k photos [mostly Nikon raw], 6k videos). When indexing photos in place on my NAS, Photoprism is able to fully utilize my hardware (CPU runs at 100% the entire time) and completes the process in several hours (6-8 usually).  If I instead choose to import my library and allow Photoprism to decide folder hierarchy the import process is glacially slow (CPU is only running at ~10% utilization) . The import has been running for 36 hours and only completed ~25k photos. 

For both tests these were fresh Photoprism setups.  I dropped the DB and recreated before each test. 

The source folder in both cases is on an NFS4 mount connected via 10GbE to a high end Synololgy NAS (8 spindles with 2TiB of NVMe cache).  When conducting throughput testing I regularly see 400-500MiB/sec to the mount point from the Photoprism box. 

I've installed MariaDB directly on the box (no docker).  

Questions: 

1) What are the actual differences between "import" and "index" beyond the obvious of the file copy/move?
2) I see quite a bit of churn on MariaDB.  Is it possible this is caused by some sort of DB locking? Is there a MariaDB tuning/setup guide I should be following?

Any ideas to help me root cause this?

**Comment by lastzero** [comment, marked as answer]
Think I can stop reading at NFS4: When import and originals folders are not on the same file system (or mounted like this via Docker), PhotoPrism has to copy files and can't just move them. Even with the fastest server and network, this is significantly slower than just indexing originals without moving them.

**Reply by jvallery** [reply]
Can you address question 1 or point me to a doc or the code? I'm curious what you're actually doing during import.  I assume this also has a dedupe step which includes a checksum calculation?  Is that done in the index case?

**Reply by lastzero** [reply]
I can point you to the exact line of code that does the copying - it happens when rename returns an error, so ultimately this is a Unix / Linux / Filesystem limitation if you like:

https://github.com/photoprism/photoprism/blob/develop/pkg/fs/move.go#L19

**Reply by lastzero** [reply]
When indexing, we don't rename, copy or move anything. So this may technically only happen while importing, where you move files from A to B.

**Reply by jvallery** [reply]
My Go knowledge is admittedly limited, but I'm curious if there is a file copy approach that would leverage a local memory buffer to improve performance and smooth out the IO.

**Reply by lastzero** [reply]
Wouldn't it be MUCH easier and effective to use / mount the same file system for import & originals?

For example, you may set `PHOTOPRISM_ORIGINALS_PATH` to `/nfs/originals` and `PHOTOPRISM_IMPORT_PATH` to `/nfs/import` and then mount your NFS share to `/nfs` in your container.
